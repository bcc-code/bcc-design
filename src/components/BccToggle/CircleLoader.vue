<template>
  <svg fill="none" viewBox="0 0 58 58" aria-hidden="true">
    <g transform="translate(2 1)" stroke="currentColor" stroke-width="1.5" class="h-3 w-3">
      <circle cx="42.601" cy="11.462" r="5" fill-opacity="1" fill="currentColor">
        <animate
          attributeName="fill-opacity"
          begin="0s"
          dur="1.3s"
          values="1;0;0;0;0;0;0;0"
          calcMode="linear"
          repeatCount="indefinite"
        />
      </circle>
      <circle cx="49.063" cy="27.063" r="5" fill-opacity="0" fill="currentColor">
        <animate
          attributeName="fill-opacity"
          begin="0s"
          dur="1.3s"
          values="0;1;0;0;0;0;0;0"
          calcMode="linear"
          repeatCount="indefinite"
        />
      </circle>
      <circle cx="42.601" cy="42.663" r="5" fill-opacity="0" fill="currentColor">
        <animate
          attributeName="fill-opacity"
          begin="0s"
          dur="1.3s"
          values="0;0;1;0;0;0;0;0"
          calcMode="linear"
          repeatCount="indefinite"
        />
      </circle>
      <circle cx="27" cy="49.125" r="5" fill-opacity="0" fill="currentColor">
        <animate
          attributeName="fill-opacity"
          begin="0s"
          dur="1.3s"
          values="0;0;0;1;0;0;0;0"
          calcMode="linear"
          repeatCount="indefinite"
        />
      </circle>
      <circle cx="11.399" cy="42.663" r="5" fill-opacity="0" fill="currentColor">
        <animate
          attributeName="fill-opacity"
          begin="0s"
          dur="1.3s"
          values="0;0;0;0;1;0;0;0"
          calcMode="linear"
          repeatCount="indefinite"
        />
      </circle>
      <circle cx="4.938" cy="27.063" r="5" fill-opacity="0" fill="currentColor">
        <animate
          attributeName="fill-opacity"
          begin="0s"
          dur="1.3s"
          values="0;0;0;0;0;1;0;0"
          calcMode="linear"
          repeatCount="indefinite"
        />
      </circle>
      <circle cx="11.399" cy="11.462" r="5" fill-opacity="0" fill="currentColor">
        <animate
          attributeName="fill-opacity"
          begin="0s"
          dur="1.3s"
          values="0;0;0;0;0;0;1;0"
          calcMode="linear"
          repeatCount="indefinite"
        />
      </circle>
      <circle cx="27" cy="5" r="5" fill-opacity="0" fill="currentColor">
        <animate
          attributeName="fill-opacity"
          begin="0s"
          dur="1.3s"
          values="0;0;0;0;0;0;0;1"
          calcMode="linear"
          repeatCount="indefinite"
        />
      </circle>
    </g>
  </svg>
</template>
