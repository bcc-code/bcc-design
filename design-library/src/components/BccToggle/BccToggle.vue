<script setup lang="ts">
import type { VueComponent } from "@/types";
import { CheckIcon, CloseIcon } from "@bcc-code/icons-vue";
import { computed, toRefs } from "vue";
import CircleLoader from "./CircleLoader.vue";

type Props = {
  modelValue: boolean;
  wasToggled?: boolean;
  disabled?: boolean;
  loading?: boolean;
  withIcon?: boolean;
  icon?: VueComponent;
  label?: string;
};

const props = withDefaults(defineProps<Props>(), {
  wasToggled: false,
  disabled: false,
  loading: false,
  withIcon: true,
  label: "",
});

const { modelValue } = toRefs(props);

const emit = defineEmits(["update:modelValue"]);

const toggled = computed({
  get() {
    return !!modelValue.value;
  },
  set(value: boolean) {
    emit("update:modelValue", value);
  },
});
</script>

<template>
  <label class="bcc-toggle" :class="{ 'bcc-toggle-was-toggled': wasToggled }">
    <input
      type="checkbox"
      class="bcc-toggle-input"
      :disabled="loading || disabled"
      v-model="toggled"
    />
    <span class="bcc-toggle-circle" aria-hidden="true">
      <CircleLoader class="size-3" v-if="loading" />
      <component v-else-if="icon" :is="icon" class="bcc-toggle-icon size-3" />
      <template v-else-if="withIcon">
        <CheckIcon
          class="bcc-toggle-icon hidden size-3 [.bcc-toggle-input:checked~.bcc-toggle-circle>&]:block"
        />
        <CloseIcon
          class="bcc-toggle-icon size-3 [.bcc-toggle-input:checked~.bcc-toggle-circle>&]:hidden"
        />
      </template>
    </span>
    <span v-if="label" class="bcc-toggle-label">{{ label }}</span>
  </label>
</template>
