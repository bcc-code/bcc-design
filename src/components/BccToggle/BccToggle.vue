<script setup lang="ts">
import { computed, toRefs } from "vue";
import { CheckIcon, CloseIcon } from "@bcc-code/icons-vue";
import CircleLoader from "./CircleLoader.vue";

type Props = {
  modelValue: boolean;
  wasToggled?: boolean;
  disabled?: boolean;
  loading?: boolean;
  withIcon?: boolean;
  label?: string;
};

const props = withDefaults(defineProps<Props>(), {
  wasToggled: false,
  disabled: false,
  loading: false,
  withIcon: true,
  label: "",
});

const { modelValue, wasToggled, disabled, loading } = toRefs(props);

const emits = defineEmits(["update:modelValue"]);

const toggled = computed({
  get() {
    return !!modelValue.value;
  },
  set(value: boolean) {
    emits("update:modelValue", value);
  },
});
</script>

<template>
  <label class="bcc-toggle" :class="{ 'bcc-toggle-was-toggled': wasToggled }">
    <input
      type="checkbox"
      class="bcc-toggle-input"
      :disabled="loading || disabled"
      v-model="toggled"
    />
    <span class="bcc-toggle-circle" aria-hidden="true">
      <CircleLoader class="h-3 w-3" v-if="loading" />
      <template v-else-if="withIcon">
        <CheckIcon class="hidden h-3 w-3 [.bcc-toggle-input:checked~.bcc-toggle-circle>&]:block" />
        <CloseIcon class="h-3 w-3 [.bcc-toggle-input:checked~.bcc-toggle-circle>&]:hidden" />
      </template>
    </span>
    <span v-if="label" class="bcc-toggle-label">{{ label }}</span>
  </label>
</template>
