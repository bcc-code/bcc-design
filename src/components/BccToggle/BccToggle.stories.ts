import BccToggle from "./BccToggle.vue";

import type { Meta, StoryFn } from "@storybook/vue3";

export default {
  title: "Components/BccToggle",
  component: BccToggle,
  argTypes: {},
} as Meta<typeof BccToggle>;

const Template: StoryFn<typeof BccToggle> = (args) => ({
  components: { BccToggle },
  setup() {
    return { args };
  },
  template: `
    <BccToggle v-model="args.modelValue" />
  `,
});

export const Example = Template.bind({});
Example.args = {
  modelValue: false,
  wasToggled: false,
  disabled: false,
  loading: false,
  withIcon: true,
  label: "Test label",
};

export const Basic: StoryFn<typeof BccToggle> = () => ({
  components: { BccToggle },
  template: `
    <div class="flex items-center space-x-2">
      <BccToggle :modelValue="true" />
      <BccToggle :modelValue="false" />
      <BccToggle :modelValue="false" :label="'Test label'" />
    </div>
  `,
});

/**
 * Set the `was-toggled` prop to `true` to indicate that changing the value of the toggle to off has consequences.
 */
export const WasToggled: StoryFn<typeof BccToggle> = () => ({
  components: { BccToggle },
  template: `
    <div class="flex items-center space-x-2">
      <BccToggle :modelValue="true" :was-toggled="true" />
      <BccToggle :modelValue="false" :was-toggled="true" />
      <BccToggle :modelValue="false" :was-toggled="true" :label="'Test label'" />
    </div>
  `,
});

export const Disabled: StoryFn<typeof BccToggle> = () => ({
  components: { BccToggle },
  template: `
    <div class="flex items-center space-x-2">
      <BccToggle :modelValue="true" disabled />
      <BccToggle :modelValue="false" disabled />
      <BccToggle :modelValue="false" disabled :was-toggled="true" />
      <BccToggle :modelValue="false" disabled :was-toggled="true" :label="'Test label'" />
    </div>
  `,
});

export const Loading: StoryFn<typeof BccToggle> = () => ({
  components: { BccToggle },
  template: `
    <div class="flex items-center space-x-2">
      <BccToggle :modelValue="true" loading />
      <BccToggle :modelValue="false" loading />
      <BccToggle :modelValue="false" :was-toggled="true" loading />
      <BccToggle :modelValue="false" :was-toggled="true" loading :label="'Test label'" />
    </div>
    <div class="flex items-center space-x-2 mt-4">
      <BccToggle :modelValue="true" loading disabled />
      <BccToggle :modelValue="false" loading disabled />
      <BccToggle :modelValue="false" :was-toggled="true" loading disabled />
      <BccToggle :modelValue="false" loading disabled :was-toggled="true" :label="'Test label'" />
    </div>
  `,
});
