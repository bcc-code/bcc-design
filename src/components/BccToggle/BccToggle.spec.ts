import { describe, it, expect } from "vitest";

import { mount } from "@vue/test-utils";
import BccToggle from "./BccToggle.vue";

describe("BccToggle", () => {
  it("renders a toggle in on or off state", () => {
    const wrapper = mount(BccToggle, { props: { modelValue: false } });
    expect(wrapper.html()).toMatchSnapshot();

    wrapper.setProps({ modelValue: true });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
