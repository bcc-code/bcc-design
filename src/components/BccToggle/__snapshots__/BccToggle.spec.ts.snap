// Vitest Snapshot v1

exports[`BccToggle > renders a toggle in on or off state 1`] = `"<button class=\\"relative inline-flex h-6 w-11 shrink-0 cursor-pointer rounded-full border-2 border-transparent transition-colors duration-200 ease-in-out focus:outline-none focus:ring-2 focus:ring-offset-2 bg-gray-200 focus:ring-gray-600\\" id=\\"headlessui-switch-1\\" role=\\"switch\\" type=\\"button\\" tabindex=\\"0\\" aria-checked=\\"false\\" data-headlessui-state=\\"\\"><span class=\\"sr-only\\">Toggle</span><span class=\\"translate-x-0 pointer-events-none relative inline-block h-5 w-5 transform rounded-full bg-white shadow ring-0 transition duration-200 ease-in-out\\"><span class=\\"opacity-100 duration-200 ease-in absolute inset-0 flex h-full w-full items-center justify-center transition-opacity\\" aria-hidden=\\"true\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" aria-hidden=\\"true\\" viewBox=\\"0 96 960 960\\" fill=\\"currentColor\\" class=\\"h-3 w-3 text-gray-400\\"><path d=\\"M480 623.739 272.87 830.87q-10.196 10.195-23.87 10.195-13.674 0-23.87-10.195-10.195-10.196-10.195-23.87 0-13.674 10.195-23.87L432.261 576 225.13 368.87q-10.195-10.196-10.195-23.87 0-13.674 10.195-23.87 10.196-10.195 23.87-10.195 13.674 0 23.87 10.195L480 528.261 687.13 321.13q10.196-10.195 23.87-10.195 13.674 0 23.87 10.195 10.195 10.196 10.195 23.87 0 13.674-10.195 23.87L527.739 576 734.87 783.13q10.195 10.196 10.195 23.87 0 13.674-10.195 23.87-10.196 10.195-23.87 10.195-13.674 0-23.87-10.195L480 623.739Z\\"></path></svg></span><span class=\\"opacity-0 duration-100 ease-out absolute inset-0 flex h-full w-full items-center justify-center transition-opacity\\" aria-hidden=\\"true\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" aria-hidden=\\"true\\" viewBox=\\"0 96 960 960\\" fill=\\"currentColor\\" class=\\"h-3 w-3 text-silver-tree-600\\"><path d=\\"M378 802.065q-6.717 0-12.554-2.358-5.837-2.359-11.316-7.837l-181-181q-10.195-10.196-10.195-24.87 0-14.674 10.195-24.87 10.196-10.195 23.99-10.195 13.793 0 24.228 10.195L378 717.782 737.891 358.13q10.196-10.195 24.37-10.315 14.174-.119 24.369 10.315 10.196 10.196 10.196 24.49 0 14.293-10.196 24.489L401.87 791.87q-5.479 5.478-11.316 7.837-5.837 2.358-12.554 2.358Z\\"></path></svg></span></span></button>"`;

exports[`BccToggle > renders a toggle in on or off state 2`] = `"<button class=\\"relative inline-flex h-6 w-11 shrink-0 cursor-pointer rounded-full border-2 border-transparent transition-colors duration-200 ease-in-out focus:outline-none focus:ring-2 focus:ring-offset-2 bg-gray-200 focus:ring-gray-600\\" id=\\"headlessui-switch-1\\" role=\\"switch\\" type=\\"button\\" tabindex=\\"0\\" aria-checked=\\"false\\" data-headlessui-state=\\"\\"><span class=\\"sr-only\\">Toggle</span><span class=\\"translate-x-0 pointer-events-none relative inline-block h-5 w-5 transform rounded-full bg-white shadow ring-0 transition duration-200 ease-in-out\\"><span class=\\"opacity-100 duration-200 ease-in absolute inset-0 flex h-full w-full items-center justify-center transition-opacity\\" aria-hidden=\\"true\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" aria-hidden=\\"true\\" viewBox=\\"0 96 960 960\\" fill=\\"currentColor\\" class=\\"h-3 w-3 text-gray-400\\"><path d=\\"M480 623.739 272.87 830.87q-10.196 10.195-23.87 10.195-13.674 0-23.87-10.195-10.195-10.196-10.195-23.87 0-13.674 10.195-23.87L432.261 576 225.13 368.87q-10.195-10.196-10.195-23.87 0-13.674 10.195-23.87 10.196-10.195 23.87-10.195 13.674 0 23.87 10.195L480 528.261 687.13 321.13q10.196-10.195 23.87-10.195 13.674 0 23.87 10.195 10.195 10.196 10.195 23.87 0 13.674-10.195 23.87L527.739 576 734.87 783.13q10.195 10.196 10.195 23.87 0 13.674-10.195 23.87-10.196 10.195-23.87 10.195-13.674 0-23.87-10.195L480 623.739Z\\"></path></svg></span><span class=\\"opacity-0 duration-100 ease-out absolute inset-0 flex h-full w-full items-center justify-center transition-opacity\\" aria-hidden=\\"true\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" aria-hidden=\\"true\\" viewBox=\\"0 96 960 960\\" fill=\\"currentColor\\" class=\\"h-3 w-3 text-silver-tree-600\\"><path d=\\"M378 802.065q-6.717 0-12.554-2.358-5.837-2.359-11.316-7.837l-181-181q-10.195-10.196-10.195-24.87 0-14.674 10.195-24.87 10.196-10.195 23.99-10.195 13.793 0 24.228 10.195L378 717.782 737.891 358.13q10.196-10.195 24.37-10.315 14.174-.119 24.369 10.315 10.196 10.196 10.196 24.49 0 14.293-10.196 24.489L401.87 791.87q-5.479 5.478-11.316 7.837-5.837 2.358-12.554 2.358Z\\"></path></svg></span></span></button>"`;
